### ISSUE TITLE
Softmax with dynamic shape gives error

### ISSUE THREAD

**Issue by oyazdanb** [iree-org/iree#22265]
## Summary
Compilation fails in the LLVMGPUVectorDistribute pass when applying softmax to tensors with dynamic dimensions. The failure occurs in gpt-oss model with streaming attention using sink tokens.

## Reproduction
Python code that trigegers the issue
```python
sink = sink.to(q.dtype)
sink = sink.reshape(1, -1, 1, 1).expand(bs, -1, n_tokens, 1)
attn_weights = ops.cat([attn_weights, sink], dim=-1)
attn_weights = ops.softmax(attn_weights, dim=-1)[..., :-1]
```
This creates a softmax operation on shape `[4, 64, 1, ?]` where `?` is dynamic.

### MLIR Files
- Toy model IR: [toy model MLIR](https://gist.github.com/oyazdanb/1173fac61c567ea8a336c3fcd199dc91)
- Full compilation error: [error log](https://gist.github.com/oyazdanb/3e15326e67dda154fc0c63d8e4baf67f)
- IR before failing pass: [before_vector_distribute.log](https://gist.github.com/oyazdanb/f236c96303c3ffe98476127d9466efe2)

### Compilation Command
```bash
iree-compile \
  gpt-oss-toy_model.mlir \
  --iree-hal-target-device=hip \
  --iree-hip-target=gfx942 \
  -o output.vmfb
```

## Error Message
`  %831 = torch.aten.softmax.int %830, %int-1_360, %none_361 : !torch.vtensor<[4,64,1,?],bf16>, !torch.int, !torch.none -> !torch.vtensor<[4,64,1,?],bf16>`

## Workaround
Restructuring the code to avoid concatenation before softmax resolves the issue:

```python
# Compute log-sum-exp for numerical stability
max_attn_weights = torch.max(attn_weights, dim=-1, keepdim=True)[0]
lse = max_attn_weights + ops.log(
    torch.sum(torch.exp(attn_weights - max_attn_weights), dim=-1, keepdim=True)
)
lse = lse.squeeze(-1)

# Apply softmax on original dimension (no concatenation)
attn_weights = ops.softmax(attn_weights, dim=-1)

# Use sink for post-softmax weighting
sink_expanded = sink.view(1, -1, 1)
alpha = ops.sigmoid(lse - sink_expanded)
result = ops.matmul(attn_weights, v) * alpha.unsqueeze(-1)
```

This avoids the error by:
1. Not creating the `dimension + 1` pattern from concatenation
2. Applying softmax to the original dynamic dimension
3. Using sink as a multiplicative weight after softmax

## Additional Investigation
To see more details about the IR state before the failing pass:
```bash
iree-compile \
  gpt-oss-toy_model.mlir \
  --iree-hal-target-device=hip \
  --iree-hip-target=gfx942 \
  --mlir-print-ir-before=iree-llvmgpu-vector-distribute \
  --mlir-disable-threading \
  -o /tmp/test.vmfb 2>&1 | tee before_vector_distribute.log
```

The failing dispatch `decode_bs4$async_dispatch_32_softmax_4x8xDxf32_dispatch_tensor_store` is shown in lines 488-588 of [this log](https://gist.github.com/oyazdanb/f236c96303c3ffe98476127d9466efe2).

**Comment by MaheshRavishankar** [comment]
I am leaving some notes here on what I found. I think we need to step back and see what is it that we are actually doing here to decide how to go about fixing this in the compiler. The problematic dispatch here is 

```
  %105 = flow.dispatch.region -> (tensor<4x8x?xf32>{%99}) {
    %238 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d1, d2, d0)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%101 : tensor<?x4x8x\
f32>) outs(%102 : tensor<4x8x?xf32>) {
    ^bb0(%in: f32, %out: f32):
      linalg.yield %in : f32
    } -> tensor<4x8x?xf32>
    %239 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%238 : tensor<4x8x?xf32\
>) outs(%104 : tensor<4x8xf32>) {
    ^bb0(%in: f32, %out: f32):
      %241 = arith.maximumf %in, %out : f32
      linalg.yield %241 : f32
    } -> tensor<4x8xf32>
    %240 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel",\
 "parallel"]} ins(%101, %239 : tensor<?x4x8xf32>, tensor<4x8xf32>) outs(%102 : tensor<4x8x?xf32>) {
    ^bb0(%in: f32, %in_128: f32, %out: f32):
      %241 = arith.subf %in, %in_128 : f32
      %242 = math.exp %241 : f32
      linalg.yield %242 : f32
    } -> tensor<4x8x?xf32>
    flow.return %240 : tensor<4x8x?xf32>
  }
```

The reduction diimension here is dynamic (see the `linalg.generic` with `reduction` iterator type here). This is going to take some doing to support in the compiler, but just to dig a bit deeper, the reduction dimension is `%99`  which is 

```
  %52 = util.assume.int %48<umin = 1, umax = 3> : index
  %87 = arith.muli %52, %c32 overflow<nsw> : index
  %99 = arith.addi %87, %c1 : index
```



This is a small reduction dimension is that always going to be the case? If the reduction dimension truly dynamic, then it needs some divisibility information. Depending on the "nature" of the dynamic support needed, we will have to teach the compiler different ways of handling this. 

### Note on creating simple repros ###

This could have very easily be repro-ed by using a single softmax operation with dynamic dimensions. Having a repro that is the full model is expensive and it will take time for resolution.  In this case

```
sink = sink.to(q.dtype)
sink = sink.reshape(1, -1, 1, 1).expand(bs, -1, n_tokens, 1)
attn_weights = ops.cat([attn_weights, sink], dim=-1)
attn_weights = ops.softmax(attn_weights, dim=-1)[..., :-1]
```

was a really good starting point. All this needed was an equivalent MLIR function that represents these operations. We really need to get down to ways in creating small MLIR repros for model level issues.

**Comment by MaheshRavishankar** [comment]
@oyazdanb from my perspective we need to step back and see what we want to do here. This is a long thread of work, and we need to take a wholistic view of what is it we are trying to achieve.

**Comment by oyazdanb** [comment]
> [@oyazdanb](https://github.com/oyazdanb) from my perspective we need to step back and see what we want to do here. This is a long thread of work, and we need to take a wholistic view of what is it we are trying to achieve.

Thanks for the guidance. we'll phase out the sink path as-is, and the current workaround covers our needs. I was only reporting the issue we encountered. If it would help the compiler team later, I can reproduce a minimal MLIR repro for the dynamic softmax reduction.

**Comment by MaheshRavishankar** [comment]
> > [@oyazdanb](https://github.com/oyazdanb) from my perspective we need to step back and see what we want to do here. This is a long thread of work, and we need to take a wholistic view of what is it we are trying to achieve.
> 
> Thanks for the guidance. we'll phase out the sink path as-is, and the current workaround covers our needs. I was only reporting the issue we encountered. If it would help the compiler team later, I can reproduce a minimal MLIR repro for the dynamic softmax reduction.

There is no need to create a smaller repro . https://github.com/iree-org/iree/issues/22265#issuecomment-3433598778 is good enough starting point for any future compiler work on this.

**Comment by Groverkss** [comment]
Cannot repro on TOM. We have invested some work in improving dynamic reductions which should have made this better.
